gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -387,103 +390,134 @@ TEST(GaussianMixtureFactor, GaussianMixtureModel2) {
+    bool add_second_measurement = false) {
+  HybridBayesNet hbn;
+
+  // Add measurement model p(z0 | x0)
+  addMeasurement(hbn, Z(0), X(0), 3.0);
 
+  // Optionally add second measurement model p(z1 | x1)
   if (add_second_measurement) {
-    // Add second measurement
-    auto p_z1 = std::make_shared<GaussianConditional>(
-        z1, Vector1(0.0), -I_1x1, x1, I_1x1, measurement_model);
-    hbn.push_back(p_z1);
+    addMeasurement(hbn, Z(1), X(1), 3.0);
   }
 
+  // Add hybrid motion model
+  hbn.push_back(hybridMotionModel);
+
   // Discrete uniform prior.
-  auto p_m1 = std::make_shared<DiscreteConditional>(m1, "0.5/0.5");
-  hbn.push_back(p_m1);
+  hbn.emplace_shared<DiscreteConditional>(m1, "50/50");
 
   return hbn;
 }
 
+/// Approximate the discrete marginal P(m1) using importance sampling
+std::pair<double, double> approximateDiscreteMarginal(
+    const HybridBayesNet& hbn,
+    const GaussianMixture::shared_ptr& hybridMotionModel,
+    const VectorValues& given, size_t N = 100000) {
+  /// Create importance sampling network q(x0,x1,m) = p(x1|x0,m1) q(x0) P(m1),
+  /// using q(x0) = N(z0, sigmaQ) to sample x0.
+  HybridBayesNet q;
+  q.push_back(hybridMotionModel);  // Add hybrid motion model
+  q.emplace_shared<GaussianConditional>(GaussianConditional::FromMeanAndStddev(
+      X(0), given.at(Z(0)), /* sigmaQ = */ 3.0));  // Add proposal q(x0) for x0
+  q.emplace_shared<DiscreteConditional>(m1, "50/50");  // Discrete prior.
+
+  // Do importance sampling
+  double w0 = 0.0, w1 = 0.0;
+  std::mt19937_64 rng(42);
+  for (int i = 0; i < N; i++) {
+    HybridValues sample = q.sample(&rng);
+    sample.insert(given);
+    double weight = hbn.evaluate(sample) / q.evaluate(sample);
+    (sample.atDiscrete(M(1)) == 0) ? w0 += weight : w1 += weight;
+  }
+  double pm1 = w1 / (w0 + w1);
+  std::cout << "p(m0) = " << 100 * (1.0 - pm1) << std::endl;
+  std::cout << "p(m1) = " << 100 * pm1 << std::endl;
+  return {1.0 - pm1, pm1};
+}
+
 }  // namespace test_two_state_estimation
 
 /* ************************************************************************* */
 /**
- * Test a model P(z0|x0)P(x1|x0,m1)P(z1|x1)P(m1).
+ * Test a model p(z0|x0)p(z1|x1)p(x1|x0,m1)P(m1).
  *
- * P(f01|x1,x0,m1) has different means and same covariance.
+ * p(x1|x0,m1) has mode-dependent mean but same covariance.
  *
- * Converting to a factor graph gives us
- * ϕ(x0)ϕ(x1,x0,m1)ϕ(x1)P(m1)
+ * Converting to a factor graph gives us ϕ(x0;z0)ϕ(x1;z1)ϕ(x1,x0,m1)P(m1)
  *
- * If we only have a measurement on z0, then
- * the probability of m1 should be 0.5/0.5.
+ * If we only have a measurement on x0, then
+ * the posterior probability of m1 should be 0.5/0.5.
  * Getting a measurement on z1 gives use more information.
  */
 TEST(GaussianMixtureFactor, TwoStateModel) {
   using namespace test_two_state_estimation;
 
   double mu0 = 1.0, mu1 = 3.0;
-  double sigma = 2.0;
-
-  DiscreteKey m1(M(1), 2);
-  Key z0 = Z(0), z1 = Z(1);
+  double sigma = 0.5;
```

Comment:
Any reason we changed the noise model value here?

Reply:
Wanted to make posterior a bit more skewed.